### PR TITLE
Fix next keyboard into a ListView and TableView iOS

### LINKED ISSUE #13176: [iOS] Pressing next into a ListView or TableView goes to last element
### Description

The code from [this PR](https://github.com/dotnet/maui/pull/13174) allows the user to click the "Next" button on the iOS Keyboard (or "enter" on physical keyboard using MacCatalyst) to automatically focus the next UITextField or UITextView (Editor and Entry in Maui). 

However, when you press the "Next" button to go into a ListView or TableView, you get sent to the last entry as you can see in the video below:

[// Insert video here](https://user-images.githubusercontent.com/50846373/217344545-0af499bc-08f4-429c-ad46-40272b046f5c.mov)

It appears that the logical tree reverses the order of the elements in the ListView and TableView and because the Next Navigation uses the logical ordering. Running a slightly different project below, you can see that when we run the maui project on VSMac on MacCatalyst and then attach a debugger inside Xcode and inspect the hierarchy tree, the TableView's entries are backwards. 

### Steps to see the process in Xcode
1. Open Xcode
2. Run the Maui project on MacCatalyst inside VSMac
3. On Xcode, click the `Debug->Attach to Process->{find the maui process and click}` dropdown on the top mac toolbar
4. Press the Debug Spray can icon on the left-side panel
image
5. Press the Debug View Hierarchy button (three stacked panels icon) on the bottom bar
6. Observe the view hierarchy on the left panel

MicrosoftTeams-image (1)

From searching inside the [ListViewRenderer](https://github.com/dotnet/maui/blob/main/src/Compatibility/Core/src/iOS/Renderers/ListViewRenderer.cs) (when observing similar behavior in the ListView), it seems that elements are in the correct order inside the DataSource so it appears that somewhere else, things are getting reversed. 

### Steps to Reproduce

1. May need to be inside the Maui repo main branch depending on this PR: https://github.com/dotnet/maui/pull/13174 (should work to reproduce this inside Maui main branch in the Sandbox)
2. Add the below xaml
3. Deploy to an iOS device or use the iOS simulator
4. Click the first entry that has the text "Initial Entry"
7. Press the "Next" button on the iOS software keyboard
8. Observe that you go into the last textfield of the tableview "E3" instead of the expected "E1"
9. Notice similar behavior inside a ListView as well. (CollectionView works as expected though)

```xaml
<VerticalStackLayout
            Spacing="25"
            Padding="30,0"
            VerticalOptions="Center">

            <Entry Text="Initial Entry" ReturnType="Next" />

            <Label Text="TableView" VerticalOptions="End"/>

            <TableView>
                <TableRoot>
                    <TableSection>
                        <EntryCell Label="EntryCell1" Text="E1"/>
                        <EntryCell Label="EntryCell2" Text="E2"/>
                        <EntryCell Label="EntryCell3" Text="E3"/>
                    </TableSection>
                </TableRoot>
            </TableView>
</VerticalStackLayout>
```

### Link to public reproduction project repository

https://github.com/tj-devel709/MauiRepros/tree/main/NextListTableViewIssueRepro

### Version with bug

7.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, macOS

### Affected platform versions

iOS 16.2 + MacCatalyst 16.2

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by tj-devel709**
### Description of Change
On iOS, pressing the next keyboard button should bring us into the first element of a ListView or TableView. However, in the logical ordering, ListView and TableView cells are out of order. By using the TableView.VisibleCells, we can reorder as needed and allow the ordering of the elements to be as we expect.

I did not see a ListView or TableView stub in the src/Core/test directory, but I created a video demo here:

https://user-images.githubusercontent.com/50846373/224452298-9f229b31-edd1-4dda-bb5f-c6a9c7cd9fac.mov

### Issues Fixed

Fixes #13176

### CODE REVIEW

File: src/Core/src/Platform/iOS/ViewExtensions.cs [hunk 1]
```diff
@@ -776,6 +787,32 @@ internal static void UpdateLayerBorder(this CoreAnimation.CALayer layer, IButton
...
+		{
+			if (isValidType(view))
+			{
+				var index = view.FindTableViewCellIndex(table);
+
+				return index == -1 ? null : table.FindNextView(index + 1, isValidType);
+			}
+
+			else
+				return table.FindNextView(0, isValidType);
```
Comment (tj-devel709) [inline src/Core/src/Platform/iOS/ViewExtensions.cs:800]: Double check if we need this else?

File: src/Core/src/Platform/iOS/ViewExtensions.cs [hunk 2]
```diff
@@ -728,6 +728,35 @@ internal static void UpdateLayerBorder(this CoreAnimation.CALayer layer, IButton
 			return null;
 		}
 
+		internal static T? FindResponder<T>(this UIViewController controller) where T : UIViewController
```
Comment (mandel-macaque) [inline src/Core/src/Platform/iOS/ViewExtensions.cs:731]: +1 on the use of generics, well done.

**Review by mandel-macaque — APPROVED** [review]

File: src/Core/src/Platform/iOS/ViewExtensions.cs [hunk 3]
```diff
@@ -784,5 +848,20 @@ internal static void ChangeFocusedView(this UIView view, UIView? newView)
 			else
 				newView.BecomeFirstResponder();
 		}
+
+		internal static UIView? SetContainerView(this UIView? startingPoint)
```
Comment (PureWeen) [inline src/Core/src/Platform/iOS/ViewExtensions.cs:852]:
```suggestion
		internal static UIView? GetContainerView(this UIView? startingPoint)
```

?

Comment (tj-devel709) [reply src/Core/src/Platform/iOS/ViewExtensions.cs]: Fixed!

**Review by PureWeen — CHANGES_REQUESTED** [review]
Can we add tests to this?

File: src/Controls/tests/DeviceTests/Elements/Entry/EntryTests.iOS.cs [hunk 4]
```diff
@@ -36,5 +42,116 @@ int GetPlatformSelectionLength(EntryHandler entryHandler)
...
+			layout.HeightRequest = 1000;
+			layout.WidthRequest = 300;
+
+			await InvokeOnMainThreadAsync(async () =>
+			{
+				var contentViewHandler = CreateHandler<LayoutHandler>(layout);
+				await contentViewHandler.PlatformView.AttachAndRun(() =>
+				{
+					KeyboardAutoManager.GoToNextResponderOrResign(entry1.ToPlatform(), customSuperView: entry1.ToPlatform().Superview);
+					var firstResponder = layout.ToPlatform().FindFirstResponder();
```
Comment (PureWeen) [inline src/Controls/tests/DeviceTests/Elements/Entry/EntryTests.iOS.cs:149]: This is probably fine. If we start seeing random failures from these tests we should move them all into a collection. As long as this block of code runs uninterrupted we should be fine.

**Review by PureWeen — APPROVED** [review]

**Pull request merged**